### DISCUSSION TITLE
The language specific sets of attributes are missing items

### DISCUSSION THREAD

**Discussion by madman-81** [Laravel-Lang/common#26] [Q&A] [answered]
- Laravel-Lang Version: 10.3
- PHP Version: 7.4

### Description:
The list of attributes in `source\validation-attributes.php` and `locales\{lang}\validation-attributes.php` are not identical. The missing keys aren't in `excludes\{lang}.php` either. 

A few examples (not complete): 
DE: missing `body`, `message`, `photo`, `price`, `role`, `subject` and `terms` but has `available`, `current_password`, `size` extra
NL: missing `body`, `photo`, `role` and `terms` but has `amount`, `available`, `teacher`, `street` and others extra

I would at least expect all attributes in `source\validation-attributes.php` to be present in the language specific file. The extra attributes are understandable and may serve a purpose. However, they make it more difficult to check if all language files are complete. If they aren't used in official supported projects (any more), it might be better to remove them.

Note: also Publisher generates different sets of attributes per language.

**Comment by andrey-helldar** [comment]
By default, Laravel does not have a set of attribute names for translation and we have placed them in `validation-attributes.php` files.

At the moment, one of the problems is that the keys are not identical in these files.

Since this list is unofficial and can be large, I suggest moving all keys from the `validation-attributes.php` files to an external project such as `laravel-lang/attributes` and keeping the keys up to date.

@Laravel-Lang/laravel-lang, what do you think?

**Reply by mohamedsabil83** [reply]
I guess it's ok for identical only as the attributes will be different from one project to another except the base ones like **name, email, password, ...etc**

**Reply by andrey-helldar** [reply]
With the current version of storing different attribute names, I also do not see a problem.

But if we support the synchronization of attributes between all localizations, then I would recommend moving them to a separate project. Fortunately, the our [publisher](https://github.com/Laravel-Lang/publisher) already, one might say, knows how to work with him :)

**Comment by madman-81** [comment]
I'm not fully aware of how all the internals of this project work, but from what I've seen there seems to be a system where new lines to the en.json are automatically added to all the {lang}.json files. Can't we use this mechanism for attributes (and perhaps the other files) as well? Or is that already in place? 

The advantage of this would be that all locale files work with the same system and that the community can add missing keys/attributes and translations. The challenge here will be maintaining the master set: who is going to decide which keys are added and which not.

**Reply by andrey-helldar** [reply]
The json file in the locales folder is populated from the key-value chain of other json files taken together, thus avoiding duplication of translation keys.

The separation is made in order to exclude keys that are not related to the project when installing translations through the publisher. For example, if you are not using Laravel Nova, then by installing updates through a publisher, Nova keys will not be installed in your application.

Also, this file does not include multi-level elements, including validation attribute names.

Since this list is not official, in this project we are not strict in their compliance. But we see the functionality of their use. Therefore, I made a proposal that our team consider the option of transferring keys to an external project. I can easily do this, but I need confirmation from the team.

**Comment by andrey-helldar** [comment, marked as answer]
I thought it over carefully and brought these keys to an external project - [`laravel-lang/attributes`](https://github.com/Laravel-Lang/attributes). A big plus of the new project is that it has no restrictions on the number of added keys 😎

Soon we will publish it in packagist and it will be available for installation via composer.
